### PR TITLE
Show examples of "manual" parameter binding

### LINKED ISSUE #25081: Expand section on parameter binding to show how to do it by hand
To help users appreciate parameter binding, we should have a section comparing common binding scenarios:
- Reading from route parameter
- Reading from query parameter
- Reading from header
- Reading JSON from the body

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: b09824ab-b051-75a1-13b7-b0e281cc46e0
* Version Independent ID: b09824ab-b051-75a1-13b7-b0e281cc46e0
* Content: [Minimal APIs overview](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/minimal-apis?view=aspnetcore-6.0)
* Content Source: [aspnetcore/fundamentals/minimal-apis.md](https://github.com/dotnet/AspNetCore.Docs/blob/main/aspnetcore/fundamentals/minimal-apis.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-fundamentals**
* GitHub Login: @Rick-Anderson
* Microsoft Alias: **riande**

### PR DESCRIPTION
**Pull request by serpent5**
Fixes #25081.

[Internal Preview](https://review.docs.microsoft.com/en-us/aspnet/core/fundamentals/minimal-apis?view=aspnetcore-6.0&branch=pr-en-us-25142#parameter-binding).

### CODE REVIEW

**Comment by serpent5** [comment]
@davidfowl I have a couple of questions for you here, if you have time:

1. Should the code example perform type conversion?
1. Should we include an example of using `GetRequiredService`? I know this wasn't in the linked issue, but it's a one-liner we could add while we're here, if that's desirable.

File: aspnetcore/fundamentals/minimal-apis/samples/WebMinAPIs/Snippets/Program.cs [hunk 1]
```diff
@@ -0,0 +1,27 @@
+﻿namespace WebMinAPIs.Snippets;
+
+public static class Program
```
Comment (davidfowl) [inline aspnetcore/fundamentals/minimal-apis/samples/WebMinAPIs/Snippets/Program.cs:3]: Why is there a program main?

File: aspnetcore/fundamentals/minimal-apis/samples/WebMinAPIs/Program.cs [hunk 2]
```diff
@@ -420,32 +420,30 @@ public static string Hello()
...
 #endregion
 
-#elif OV
-#region snippet_ov
+#elif PBP
+#region snippet_pbp
 var builder = WebApplication.CreateBuilder(args);
 
-// Added as service
-builder.Services.AddSingleton<Service>();
```
Comment (davidfowl) [inline aspnetcore/fundamentals/minimal-apis/samples/WebMinAPIs/Program.cs:442]: Why this change?

File: aspnetcore/fundamentals/minimal-apis/samples/WebMinAPIs/Snippets/Program.cs (hunk 1 above)
Comment (serpent5) [reply aspnetcore/fundamentals/minimal-apis/samples/WebMinAPIs/Snippets/Program.cs:3]: It's just the class, without a `Main` method. It's a way of including code snippets in the article that always compiles, but isn't really part of the sample. I've used a `Program` class here because it shows where the code would be, as I can't put top level statements outside of the root-level `Program.cs`.

File: aspnetcore/fundamentals/minimal-apis/samples/WebMinAPIs/Program.cs (hunk 2 above)
Comment (serpent5) [reply aspnetcore/fundamentals/minimal-apis/samples/WebMinAPIs/Program.cs:442]: The example right above this one for GET now shows resolving a service from DI, so I removed it here to shorten the example and save the repetition.

**Comment by Rick-Anderson** [comment]
@davidfowl @halter73  can you review the [Internal Preview](https://review.docs.microsoft.com/en-us/aspnet/core/fundamentals/minimal-apis?view=aspnetcore-6.0&branch=pr-en-us-25142#parameter-binding) and answer these questions:

> @davidfowl I have a couple of questions for you here, if you have time:
> 
> 1. Should the code example perform type conversion?
> 2. Should we include an example of using `GetRequiredService`? I know this wasn't in the linked issue, but it's a one-liner we could add while we're here, if that's desirable.

Did @serpent5 answer your questions?

**Comment by Rick-Anderson** [comment]
Arg
image

cc @guardrex

**Comment by Rick-Anderson** [comment]
> 
> **UPDATE**: I ran thru the open PRs and fixed them all up.

I avoid that route. If the PR has significant local changes _**OR**_ the owner doesn't notice you pushed changes and then they make significant changes and push, that can lead to more significant merge conflicts.

The real answer is to get the global file working.

File: aspnetcore/fundamentals/minimal-apis.md [hunk 3]
```diff
@@ -407,32 +407,31 @@ Supported binding sources:
...
 
-**NOTE**: Binding from forms is ***not*** natively supported in .NET 6.
+> [!NOTE]
+> Binding from form values is ***not*** natively supported in .NET 6.
 
-### GET, HEAD, OPTIONS, DELETE
+The following example GET route handler uses some of these parameter binding sources:
 
-The HTTP methods `GET`, `HEAD`, `OPTIONS`, and `DELETE` don't implicitly bind from body. Bind explicitly with  [`[FromBody]`](xref:Microsoft.AspNetCore.Mvc.FromBodyAttribute) or read from the request body. All other binding sources are supported.
+[!code-csharp[](minimal-apis/samples/WebMinAPIs/Program.cs?name=snippet_pbg&highlight=8-11)]
```
Comment (davidfowl) [inline aspnetcore/fundamentals/minimal-apis.md:415]: Why this change?

File: aspnetcore/fundamentals/minimal-apis/samples/WebMinAPIs/Snippets/Program.cs [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+            // ...
+        });
+
+        app.MapPost("/", async (HttpContext context) =>
+        {
+            var person = await context.Request.ReadFromJsonAsync<Person>();
+
+            // ...
+        });
```
Comment (davidfowl) [inline aspnetcore/fundamentals/minimal-apis/samples/WebMinAPIs/Snippets/Program.cs:22]: The only change I would make is injecting the HttpRequest instead of the HttpContext

**Review by davidfowl — APPROVED** [review]
One small nit and this is good to go.

File: aspnetcore/fundamentals/minimal-apis/samples/WebMinAPIs/Snippets/Program.cs [hunk 5]
```diff
@@ -5,18 +5,18 @@ public static class Program
     public static void ManualRequestBinding(WebApplication app)
     {
         // <snippet_ManualRequestBinding>
-        app.MapGet("/{id}", (HttpContext context) =>
+        app.MapGet("/{id}", (HttpRequest request) =>
         {
-            var id = context.Request.RouteValues["id"];
-            var page = context.Request.Query["page"];
-            var customHeader = context.Request.Headers["X-CUSTOM-HEADER"];
```
Comment (davidfowl) [inline aspnetcore/fundamentals/minimal-apis/samples/WebMinAPIs/Snippets/Program.cs]: Do we have an example anywhere on how to read cookies?

Comment (serpent5) [reply aspnetcore/fundamentals/minimal-apis/samples/WebMinAPIs/Snippets/Program.cs]: I can't find anything in the docset.

@Rick-Anderson I'd expect there to be something [here](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/app-state?view=aspnetcore-6.0#cookies), perhaps. Is there somewhere else that covers this?
